File: app/src/main/java/com/zeyad/rxredux/screens/user/ViewModelFactory.java
package com.zeyad.rxredux.screens.user;

import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;
import android.support.annotation.NonNull;

import com.zeyad.rxredux.screens.user.detail.UserDetailVM;
import com.zeyad.rxredux.screens.user.list.UserListVM;
import com.zeyad.usecases.api.IDataService;

/**
 * @author ZIaDo on 12/13/17.
 */
public class ViewModelFactory extends ViewModelProvider.NewInstanceFactory {
    final IDataService dataService;

    public ViewModelFactory(IDataService dataService) {
        this.dataService = dataService;
    }


    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if (modelClass.isAssignableFrom(UserListVM.class)) {
            return (T) new UserListVM(dataService);
        } else if (modelClass.isAssignableFrom(UserDetailVM.class)) {
            return (T) new UserDetailVM(dataService);
        }
        throw new IllegalArgumentException("Unknown ViewModel class: " + modelClass.getSimpleName());
    }
}

File: app/src/main/java/com/zeyad/rxredux/screens/ViewModelFactory.kt
package com.zeyad.rxredux.screens

import android.arch.lifecycle.ViewModel
import android.arch.lifecycle.ViewModelProvider

import com.zeyad.rxredux.screens.user.detail.UserDetailVM
import com.zeyad.rxredux.screens.user.list.UserListVM
import com.zeyad.usecases.api.IDataService

/**
 * @author ZIaDo on 12/13/17.
 */
class ViewModelFactory(private val dataService: IDataService) : ViewModelProvider.NewInstanceFactory() {


    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(UserListVM::class.java)) {
            return UserListVM(dataService) as T
        } else if (modelClass.isAssignableFrom(UserDetailVM::class.java)) {
            return UserDetailVM(dataService) as T
        }
        throw IllegalArgumentException("Unknown ViewModel class: " + modelClass.simpleName)
    }
}
